### ISSUE TITLE
Bracketed paste in normal-mode

### ISSUE THREAD

**Issue by chrisjohnson** [neovim/neovim#8395]
- `nvim --version`: 
```
$ nvim --version
NVIM v0.2.2
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/local/Homebrew/Library/Homebrew/shims/super/clang -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -I/tmp/neovim-20180209-66903-ukwirj/neovim-0.2.2/build/config -I/tmp/neovim-20180209-66903-ukwirj/neovim-0.2.2/src -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/gettext/include -I/usr/include -I/tmp/neovim-20180209-66903-ukwirj/neovim-0.2.2/build/src/nvim/auto -I/tmp/neovim-20180209-66903-ukwirj/neovim-0.2.2/build/include
Compiled by brew@Sierra-2.local

Features: +acl +iconv +jemalloc +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.2.2_1/share/nvim"

Run :checkhealth for more info
```
- Vim (version: `8.0 (2016 Sep 12, compiled Oct  5 2017 19:19:02)`) behaves differently?
  - Yes, vim 8 behaves as expected
- Operating system/version: macOS 10.12.6
- Terminal name/version: iTerm 2 Build 3.1.6beta5
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

First, in system clipboard, copy something. Here's what I'm using: `Build 3.1.6beta5`

```
nvim -u NORC
i
asdf
<esc>
yy
cmd-v
```

### Actual behaviour
The new buffer contents become:

```ld 3.1.6beta5```

### Expected behaviour
The buffer contents should become:
```asdfBuild 3.1.6beta5```

And the user should be left in normal mode

This is the behavior exhibited when using vim 8 via `vim -u NONE`

### Detail
It seems as though the contents of my clipboard are being parsed as input, which defeats the point of bracketed paste mode. If you were to repeat the original steps up until cmd-v and instead manually type out the string `Build 3.1.6beta5` from normal mode, you end up with the "Actual behavior" results

**Comment by jaawerth** [comment]
I'm seeing similar behavior

```
NVIM v0.2.2
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.2.2/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

# Test case

  1. copy `:q` to the clipboard
  2. In normal mode, `Ctrl-shift-v` (linux, terminator) to paste into terminal

In an unmodified buffer in neovim (NORC), it quits; in vim 8, it pastes `:q` and returns to normal mode.

**Comment by justinmk** [comment]
Ah , you're pasting into normal mode. Only insert mode is supported currently.

**Comment by chrisjohnson** [comment]
Is this a reasonable place to request such a feature? Has it been requested already? It seems counter to the goals of bracketed paste mode not to support pasting in normal mode.

**Comment by jamessan** [comment]
Is this another manifestation of #3447?

**Comment by jaawerth** [comment]
@justinmk Would it make sense to update `:help bracketed-paste-mode` to clarify that it only works in insert mode until it's implemented in normal mode as well, at least until this is implemented? Minor quibble, but I'd classify this as a "bug" unless it behaves as documented ;-)

At any rate, for what it's worth I'd definitely support having this functionality if only for parity with vim8 even though I didn't even notice until this came up, being so ingrained to use the + register or paste from insert mode.... having it would be convenient though!

@jamessan Good call, I'd actually put money on it; [this comment](https://github.com/neovim/neovim/issues/3447#issuecomment-370921769) indicates it isn't an issue with `mouse=a` or `mouse=n`, which makes sense, since the mouse being disabled would, at least for many terminals, make it fall back to pasting from system clipboard.

**Comment by justinmk** [comment]
It's definitely planned and I have an old PR for it

**Comment by janlazo** [comment]
https://gist.github.com/janlazo/cea7acf2711e89a9bc4f7c79612c6e54#bracketed-paste is an incomplete list of vim patches for bracketed paste but I don't want to work on them (not familiar with `src/nvim/tui.c`). I'll update the list if I find more patches.

**Comment by wincent** [comment]
This would certainly be pretty useful. Not just for parity with Vim 8; you could also get bracketed paste working in all modes in previous Vim versions with [a little help](https://github.com/wincent/terminus).

**Comment by blueyed** [comment]
This is btw what "vim-patch:8.0.0231: bracketed paste mode is not tested" (https://github.com/vim/vim/commit/076e502199b19e6141e4c1e659ff3f21b71934e1), the oldest unimported patch is testing - specifically that it gets appended always then.
It's also noteworthy that Vim displays `^[[200~pasted]` when using `<Ctrl-v>` before it (which is useful to verify that it is working I guess).

**Comment by justinmk** [comment]
paste in normal-mode is working now in https://github.com/neovim/neovim/pull/4448 .

> Vim displays `^[[200~pasted]` when using `<Ctrl-v>` before it (which is useful to verify that it is working I guess).

wontfix. Nvim UIs including the TUI do not send implementation details to the editor.

**Comment by justinmk** [comment]
Closed by #4448.

> This would certainly be pretty useful. Not just for parity with Vim 8; 

Nvim now supports this by default, and also:

- paste is 10x-100x faster (even compard to Vim 8.1 with bracketed-paste enabled)
- allows custom behavior by overriding `vim.paste` Lua function.
